Python3 Open PR

arpitn30 opened the pull request:
Kodi code ported to Python3 with system dependencies updated accordingly

## Description
The following changes have been made to the code -
* PyString has been changed to PyUnicode and PyBytes accordingly.
* PyInt has been changed to PyLong since former is no longer supported.
* Obsolete Py_TPFLAGS have been removed or updated
* Compare has been completely replaced in favor of RichCompare
* Module Initialization has been updated
  * Py_InitModule has been replaced with PyModule_Create
  * PyModuleDef added
  * Added PyImport_AppendInittab to add modules to the import modules list
  * Updated swig/CMakeLists to add Python Binding directly to Core Library
* Changed charset from char to wchar_t due to new functions using the latter
* Replaced PyFile_AsFile, now removed with a custom function
* Updated GIL acquisition and release statements
* Updated Windows package lists and directory paths to new dependency package structure
* Updated Python Package and Library versions
* Updated Posix native files and added patches wherever necessary (Patches might need to be removed when the mirrors are updated to Python3)
* Removed outdated and added new patches to target Python directory
* Updated Makefiles, dependencies and added patches for specific platforms

## Motivation and Context
It helps the add-on developers to code in Python3

## How Has This Been Tested?
Yes, this has been tested on Jenkins and on Windows, Linux, and Android using add-ons ported to Python3

## Types of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

arpitn30 committed:
python3 implementation

* Updated all the files from old function names to new function names as
  given in pydocs - http://py3c.readthedocs.io/en/latest/reference.html
* changed char to wchar_t
* File pointer (FILE *) updated
* Updated ob_type to be called by multi layered PyObject
* Removed tp_compare since it's depreciated
* Updated definitions for InitModule
* Updated the Python3 Android Script
* Updated the GIL acquisition and release statements
* Updated module initialization from init<module> to PyInit_<module> and added the new modules to built-in library through PyImport_AppendInitTab
* Charset Conversion From char To wchar_t
* Updated Python WSGI Invoker

Rechi commented on xbmc/interfaces/python/XBPython.cpp:
```diff
@@ -45,7 +45,7 @@
 #include "interfaces/python/AddonPythonInvoker.h"
 #include "interfaces/python/PythonInvoker.h"
-
+PyThreadState* savestate;
```
shouldn't this be a meber of CPythonInvoker, XBPython or something similar

arpitn30 replied:
Yeah, it was a temporary measure until a replacement to transfer the value of threadstate from XBPython to Pythoninvoker since paxxi said it wouldn't be a good idea to use global variables.
@Rechi

arpitn30 replied:
Since the functions savestate is used in don't share a common class (neither heirarchial), the option other than gloabl variable would be to define the variable static in CPythonInvoker

Rechi commented on xbmc/interfaces/python/PythonSwig.cpp.template:
```diff
@@ -930,8 +970,17 @@ namespace PythonBindings
+        return NULL;
+    struct module_state *st = GETSTATE(module);
+
+    st->error = PyErr_NewException((char *)"${module.@name}.Error", NULL, NULL);
```
this needs to be removed, I only suggested it for testing but it doesn't do what I thought

Rechi commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -67,14 +67,14 @@ namespace PythonBindings
+    else    // If pobject is of type Bytes
     {
-      buf = PyString_AsString(pObject);
+      buf = PyUnicode_AsUTF8(pObject);
```
I get crashes with a nullptr exception here

Rechi commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -67,14 +67,14 @@ namespace PythonBindings
       if (utf8_pyString)
       {
-        buf = PyString_AsString(utf8_pyString);
+        buf = PyUnicode_AsUTF8(utf8_pyString);
```
I get crashes with a nullptr exception here

arpitn30 commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -63,18 +63,14 @@ namespace PythonBindings
-        Py_DECREF(utf8_pyString);
-        return;
-      }
+      buf = PyUnicode_AsUTF8(pObject);
```
@Rechi Can you check if this works correctly?
According to the documentation, this function should encode the PyObject as UTF-8 and return the pointer to the string so wouldn't need another function to encode.

Rechi replied:
have tested that already during getting the crash and it worked

arpitn30 committed:
Converted Virtual to Pure Virtual Functions

Rechi commented on tools/depends/target/Makefile:
```diff
@@ -95,9 +95,15 @@ libpng: $(ZLIB)
+pythonmodule-pycryptodome: python36 pythonmodule-setuptools
+pythonmodule-pil: $(ZLIB) libjpeg-turbo libpng freetype2 python36 pythonmodule-setuptools
+pythonmodule-setuptools: python36
+>>>>>>> Changed Tools/Depends MakeFiles
```
rebase error

Rechi commented on tools/depends/native/python36-native/Makefile:
```diff
@@ -31,6 +31,7 @@ $(LIBDYLIB): $(PLATFORM)
 
 .installed-$(PLATFORM): $(LIBDYLIB)
 	cd $(PLATFORM); $(MAKE) install
+	cd $(PLATFORM); $(MAKE) regen-grammar
```
this is a build command, not an install command

Rechi commented on tools/depends/target/python36/Python-3.6.2-android.patch:
```diff
@@ -40,3 +40,16 @@
       METH_VARARGS, termios_tcflush__doc__},
      {"tcflow", termios_tcflow,
+
+     
```
two empty lines are unnecessary

arpitn30 replied:
I thought it would make it easier to differentiate between files in the same patch file.

Rechi replied:
IMO the lines staring with 3 -/+ are indication enough

arpitn30 commented:
The native Python modules can't be found while running add-ons when Kodi is built through CMake. No such problem is faced when built on Windows or on Linux using Clang or GCC
@Razzeee

Memphiz commented on tools/darwin/Configurations/App-iOS.xcconfig:
```diff
@@ -0,0 +1,48 @@
+//
```
that file is gone in master and uneeded.

razzeee replied:
@arpitn30 probably a rebase error

Memphiz commented on tools/darwin/Configurations/App.xcconfig.in:
```diff
@@ -0,0 +1,31 @@
+//
```
file is gone in master

Rechi commented on cmake/modules/FindPython.cmake:
```diff
@@ -6,12 +6,12 @@
 
 if(PKG_CONFIG_FOUND)
-  pkg_check_modules(PC_PYTHON python>=2.7 QUIET)
+  pkg_check_modules(PC_PYTHON python3>=3.6 QUIET)
```
`python3>=3.4`
Ubuntu 14.04 has 3.4, 16.04 has 3.5.1 and 17.04 has 3.5.3

Rechi commented on cmake/modules/FindPython.cmake:
```diff
@@ -6,12 +6,12 @@
-find_path(PYTHON_INCLUDE_DIR NAMES Python.h PATHS ${PC_PYTHON_INCLUDE_DIRS} ${DEPENDS_PATH}/include/python2.7)
+find_program(PYTHON_EXECUTABLE python3 ONLY_CMAKE_FIND_ROOT_PATH)
+find_library(PYTHON_LIBRARY NAMES python3.6 PATHS ${PC_PYTHON_LIBDIR})
+find_path(PYTHON_INCLUDE_DIR NAMES Python.h PATHS ${PC_PYTHON_INCLUDE_DIRS} ${DEPENDS_PATH}/${ARCH}/include/python3.6)
```
remove `${ARCH}/`

arpitn30 replied:
@Rechi Wouldn't removing {ARCH} break win32 builds since the include path was changed

Rechi replied:
`DEPENDS_PATH` is only set by depends (https://github.com/xbmc/xbmc/blob/f65be03/tools/depends/target/Toolchain.cmake.in#L1).
On windows this variable is empty.
I also just tested it on windows and it works correct without `{ARCH}/`

Rechi commented on tools/depends/target/pythonmodule-pil/pillow-crosscompile.patch:
```diff
@@ -3,7 +3,7 @@
  
          _add_directory(include_dirs, "libImaging")
- 
+
```
remove the whitespace changes in the whole file

arpitn30 replied:
I didn't add them, these were already there when I patched the Makefile.pre.in again. The patch might stop working after removing them
@Rechi

Rechi replied:
you introduced the whitspace changes at 4f7fc770431f58a97c7f492b146679ea563ce19d for this file.
drop them
I wonder why this even does apply

Rechi commented on cmake/modules/FindPython.cmake:
```diff
@@ -6,12 +6,12 @@
 
 if(PKG_CONFIG_FOUND)
-  pkg_check_modules(PC_PYTHON python>=2.7 QUIET)
+  pkg_check_modules(PC_PYTHON python3>=3.5 QUIET)
```
`python3>=3.4`
for Ubuntu 14.04 which will have support until April 2019

Rechi commented on cmake/modules/FindPython.cmake:
```diff
@@ -25,7 +25,7 @@ if(KODI_DEPENDSBUILD)
   set(PYTHON_LIBRARIES ${PYTHON_LIBRARY} ${FFI_LIBRARY} ${EXPAT_LIBRARY} ${INTL_LIBRARY} ${GMP_LIBRARY} ${PYTHON_DEP_LIBRARIES})
 else()
-  find_package(PythonLibs 2.7 REQUIRED)
+  find_package(PythonLibs 3.6 REQUIRED)
```
`3.4` for some Ubuntu versions

Rechi replied:
@arpitn30 looks like you missed that

arpitn30 replied:
Oops!

arpitn30 replied:
Done!

razzeee commented:
We should probably bump the xbmc.python abi?

arpitn30 committed:
Updated Windows Dependencies

arpitn30 committed:
[depends] Native Files Updated

arpitn30 committed:
[depends] Updated Python3 Makefile and Patches

arpitn30 committed:
[depends] Updated the PythonPath in target directories

arpitn30 committed:
Added IOS and OSX Fix

arpitn30 committed:
Added Android Fix

arpitn30 committed:
Added Python Version Varaiable

wsnipex committed:
[depends] build expat-native as PIC

Rechi committed:
[depends][python3] enable blake2 module

Rechi committed:
[debian][cpack] switch to python3 depends

ronie committed:
use python3 migration addon repo

Rechi committed:
[cleanup] remove unused functions

Rechi committed:
[cleanup] remove unused parameters from executeScript function

Rechi committed:
Revert "Reusepython"

Rechi committed:
[TEMP] add python BASE_URL

Rechi committed:
[depends] add native libffi

romanvm committed:
Check GIL state before calling PyEval_AcquireLock() to avoid a deadlock

In Python 3.7 Py_Initialize() now acuires the GIL that leads to
a deadlock with the subsequent PyEval_AcquireLock() call.

This should fix xbmc/xbmc#14885.

Rechi committed:
[TEMP] add windows precompiled python lib
